### ISSUE TITLE
💅 False positive for `useHookAtTopLevel`

### ISSUE THREAD

**Issue by pajter** [biomejs/biome#7097]
### Environment information

```bash
CLI:
  Version:                      2.1.3
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.16.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/11.4.2

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  true
  Enabled rules:
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noForEach
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useIsNan
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    performance/noAccumulatingSpread
    performance/noDelete
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noDefaultExport
    style/noDescendingSpecificity
    style/noNonNullAssertion
    style/useArrayLiterals
    style/useConst
    style/useDeprecatedReason
    style/useExponentiationOperator
    style/useExportType
    style/useForOf
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useShorthandFunctionType
    style/useTemplate
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConsole
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

useHookAtTopLevel

### Playground link

https://biomejs.dev/playground/?code=ZgB1AG4AYwB0AGkAbwBuACAARwBFAFQAXwBDAE8ATQBQAE8ATgBFAE4AVAAoACkAIAB7AAoACQByAGUAdAB1AHIAbgAgAGYAdQBuAGMAdABpAG8AbgAgAE0AeQBSAGUAYQBjAHQAQwBvAG0AcABvAG4AZQBuAHQAKAApACAAewAKAAkACQBjAG8AbgBzAHQAIABbAHMAdABhAHQAZQBdACAAPQAgAHUAcwBlAFMAdABhAHQAZQAoACkAOwAKAAoACQAJAHIAZQB0AHUAcgBuACAAPABkAGkAdgA%2BAHsAcwB0AGEAdABlAH0APAAvAGQAaQB2AD4AOwAKAAkAfQA7AAoAfQAKAAoAZgB1AG4AYwB0AGkAbwBuACAAZwBlAHQAQwBvAG0AcABvAG4AZQBuAHQAKAApACAAewAKAAkAcgBlAHQAdQByAG4AIABmAHUAbgBjAHQAaQBvAG4AIABNAHkAUgBlAGEAYwB0AEMAbwBtAHAAbwBuAGUAbgB0ACgAKQAgAHsACgAJAAkAYwBvAG4AcwB0ACAAWwBzAHQAYQB0AGUAXQAgAD0AIAB1AHMAZQBTAHQAYQB0AGUAKAApADsACgAKAAkACQByAGUAdAB1AHIAbgAgADwAZABpAHYAPgB7AHMAdABhAHQAZQB9ADwALwBkAGkAdgA%2BADsACgAJAH0AOwAKAH0ACgA%3D&ruleDomains.react=all&ruleDomains.project=recommended

### Expected result

I would expect that the casing of the function name (that returns a React component) is irrelevant.

It seems that whenever the first character is uppercase, the rule triggers a false positive. In the playground, you can see that `getComponent` does not raise any errors.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by arendjr** [comment]
> It seems that whenever the first character is uppercase, the rule triggers a false positive. In the playground, you can see that `getComponent` does not raise any errors.

Unfortunately, React hook rules depend largely on heuristics to determine what is a hook and what is a component function. But I think this case can be resolved by looking at the capitalisation of the `MyReactComponent` inner function, after which the outer function should hopefully not matter anymore.

**Comment by vikkrantxx7** [comment]
> > It seems that whenever the first character is uppercase, the rule triggers a false positive. In the playground, you can see that `getComponent` does not raise any errors.
> 
> Unfortunately, React hook rules depend largely on heuristics to determine what is a hook and what is a component function. But I think this case can be resolved by looking at the capitalisation of the `MyReactComponent` inner function, after which the outer function should hopefully not matter anymore.

I think that didn't fix it.

**Comment by arendjr** [comment]
Okay, sorry, but then I may not be sure how to fix it. I guess the React team should've thought twice before relying on heuristics only to decide what's a function and what's a component...

**Comment by awreese** [comment]
I came looking for a solution to this, though my problem is entirely different. We've a considerable number of false-positives with React-Table utilities that are all named with a `use-` prefix (`usePagination`, `useRowSelect`, `useSortBy`, and so on).

I'm hoping for a configuration option or similar like `useExhaustiveDependencies` has [options](https://biomejs.dev/linter/rules/use-exhaustive-dependencies/#options). It would be nice to be able to specify a list of functions that are not actually React hooks at all.

Using a `biome-ignore` comment for each is a solution... the only option at the moment, but it would be nice to whitelist known false-positives ahead of time and not be required to sprinkle disabling comments all over the place each time they are used.

**Comment by dyc3** [comment]
@awreese just merged an `ignore` option for you. It'll be available in 2.4

**Comment by awreese** [comment]
> [@awreese](https://github.com/awreese) just merged an `ignore` option for you. It'll be available in 2.4

@dyc3  Thanks, an unexpectedly quick response! I look forward to that release. Any idea when v2.4 will be available?

**Comment by dyc3** [comment]
Unfortunately not. We usually don't give out ETAs

**Comment by ematipico** [comment]
Closing. The issues has been fixed in one of latest versions of Biome.
